### ISSUE TITLE
Class variables redefinition in children with nilable type

### ISSUE THREAD

**Issue by vladfaust** [crystal-lang/crystal#4869]
```
class Foo
  @@foo : Int32? = nil
end

class Bar < Foo
  @@foo = 42
end

Error in temp.cr:6: class variable '@@foo' of Bar is already defined as (Int32 | Nil) in Foo
  @@foo = 42
  ^
```
[Play](https://carc.in/#/r/2kq5)

It's an **unexpected** behaviour for me.

However, this works but ugly:
```
class Bar < Foo
  @@foo = 42.as(Int32?)
end
```
Needs improvement?

**Comment by MakeNowJust** [comment]
I think this is the compiler's bug so we can improve it because this can be compiled:

```crystal
class Foo
  @x : Int32?

  def x
    @x
  end
end

class Bar < Foo
  @x = 42
end

pp Foo.new.x # => nil
pp Bar.new.x # => 42
```

To fix this is done in #4871. Thank you for reporting, @vladfaust.

**Comment by vladfaust** [comment]
Yay, my first successful bug report in Crystal, I'm so happy! Thanks, @MakeNowJust
